Fix darglint issues

jmaunon opened the pull request:
## Description
Fix darglint issues. 

Closes #73
Closes #28

Status:

- [x] config
- [x] encoder
- [x] protocol
  - [x] beaver
  - [x] fss (it still raises somo erros, help needed)
  - [x] spdz
- [x] session
- [x] store
- [x] tensor
- [x] utils
-
With this PR docstring global docstring has increased as well. (It is feasible to reach even 100%)

=========== Coverage for /home/runner/work/SyMPC/SyMPC/src/sympc/ =============
----------------------------------- Summary ------------------------------------
| Name                                 |   Total |   Miss |   Cover |   Cover% |
|--------------------------------------|---------|--------|---------|----------|
| __init__.py                          |       2 |      1 |       1 |      50% |
| config/__init__.py                   |       1 |      0 |       1 |     100% |
| config/config.py                     |       3 |      0 |       3 |     100% |
| encoder/__init__.py                  |       1 |      0 |       1 |     100% |
| encoder/fp_encoder.py                |      10 |      2 |       8 |      80% |
| module/__init__.py                   |       3 |      2 |       1 |      33% |
| module/nn/__init__.py                |       1 |      1 |       0 |       0% |
| module/nn/conv.py                    |       7 |      0 |       7 |     100% |
| module/nn/functional.py              |       2 |      0 |       2 |     100% |
| module/nn/linear.py                  |       7 |      1 |       6 |      86% |
| module/nn/smpc_module.py             |       3 |      1 |       2 |      67% |
| protocol/__init__.py                 |       1 |      0 |       1 |     100% |
| protocol/protocol.py                 |       2 |      0 |       2 |     100% |
| protocol/beaver/__init__.py          |       1 |      0 |       1 |     100% |
| protocol/beaver/beaver.py            |      12 |      0 |      12 |     100% |
| protocol/fss/__init__.py             |       1 |      1 |       0 |       0% |
| protocol/fss/fss.py                  |      26 |      5 |      21 |      81% |
| protocol/spdz/__init__.py            |       1 |      0 |       1 |     100% |
| protocol/spdz/spdz.py                |       6 |      0 |       6 |     100% |
| session/__init__.py                  |       1 |      0 |       1 |     100% |
| session/session.py                   |       5 |      0 |       5 |     100% |
| session/session_manager.py           |       5 |      0 |       5 |     100% |
| store/__init__.py                    |       7 |      3 |       4 |      57% |
| store/crypto_primitive_provider.py   |       6 |      2 |       4 |      67% |
| store/crypto_store.py                |       5 |      0 |       5 |     100% |
| tensor/__init__.py                   |       1 |      0 |       1 |     100% |
| tensor/mpc_tensor.py                 |      30 |      2 |      28 |      93% |
| tensor/share_tensor.py               |      21 |      2 |      19 |      90% |
| utils/__init__.py                    |       1 |      0 |       1 |     100% |
| utils/mpc_utils.py                   |       7 |      0 |       7 |     100% |
| utils/utils.py                       |       6 |      0 |       6 |     100% |
|--------------------------------------|---------|--------|---------|----------|
| TOTAL                                |     185 |     23 |     162 |    87.6% |
---------------- RESULT: PASSED (minimum: 52.5%, actual: 87.6%) ----------------
## Affected Dependencies
I have included in pre-commit the following:
- `pydocstyle`: a static analysis tool for checking compliance with Python docstring conventions.
- `darglint`: checks whether a docstring's description matches the actual function/method implementation. Darglint does not check stylistic preferences expressed by tools in the Python Code Quality Authority (through tools such as pydocstyle). So when using Darglint, it may be a good idea to also use pydocstyle, if you want to enforce style. (For example, pydocstyle requires the short summary to be separated from other sections by a line break. Darglint makes no such check.)

Notice as well that this affect `github actions` meaning that if your code raises an error from `pydocstyle` or `darglint` , you cannot merge. At the beginning is a bit annoying, but is the best form to not to repeat this PR in a couple of months.

## How has this been tested?
- `darglint src` does not raise errors.


## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

gmuraru commented on requirements.dev.txt:
```diff
@@ -4,6 +4,7 @@ bandit==1.7.0
 black==20.8b1
 codecov==2.1.11
 darglint==1.7.0
```
Q: What is the difference between ```darglint``` and ```docformatter```?

gmuraru replied:
Is ```docformatter``` something that it is used with ```darglint```?

jmaunon replied:
@gmuraru, darglint/pydocstyle checks the documentation, docformatter formats your code.
Darglint ~ flake, docformatter ~ black.

This kind of things helps to keep the code well-formed, but at the beginning can be a bit annoying if the developer is not familiar with pep conventions. I will discuss with the `docs_team` if this should be (or not) an standard for every openmined project.

jmaunon commented:
I have updated docstring from protocols package. However there are some methods that are not totally clear for me (marked with a TODO)
@gmuraru . Could you help me and fix some missing docstring in protocol/beaver?
@LaRiffle . Could you help me and fix some missing docstring in protocol/fss?

You can push directloy to this branch. 
In the meantime I will continue with the rest of packages/modules.

gmuraru commented on .pep8speaks.yml:
```diff
@@ -0,0 +1,28 @@
+        - W504  # line break after binary operator
+        - E402  # module level import not at top of file
+        - E731  # do not assign a lambda expression, use a def
+        - C406  # Unnecessary list literal - rewrite as a dict literal.
```
Q: Shouldn't we allow this?

jmaunon replied:
Forget this. This file will be removed, I wanted to test this functionality, but must be activated at repository level, and I am not sure about the benefit.

gmuraru commented on src/sympc/config/config.py:
```diff
@@ -7,18 +7,25 @@
-        encoder_precision (int): the precision for the encoder
+    Attributes:
+        encoder_base (int): Base for the encoder.
+        encoder_precision: Precision for the encoder.
```
Q: Should we also add the type?

jmaunon replied:
Done

gmuraru commented on src/sympc/config/config.py:
```diff
@@ -7,18 +7,25 @@
+        Config can be used inside a Session (see Session) to
+        specify different options that can be used for the Fixed Precision Encoder.
+
+        Args:
```
Q: Should we have ```Args``` or ```Attributes``` or ```Arguments```?

jmaunon replied:
`Attributes` key is used for the attributes of a class.
`Args` key is used for arguments of a function or method from from a class
`Arguments` is not useed in Google format (as far as I know): see https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

gmuraru commented:
Made a PR with the Beaver doc [here](https://github.com/jmaunon/SyMPC/pull/1)

jmaunon commented:
> Made a PR with the Beaver doc [here](https://github.com/jmaunon/SyMPC/pull/1)

Thanks for your PR. I have merged and now `darglint` seems ok with this module :100:

gmuraru commented:
This should also fix #28, right?

jmaunon commented:
> This should also fix #28, right?

Yes. I am updating everything to Google docstyle

jmaunon commented:
@gmuraru , this PR is ready  for reviewing/merging. I have updated the description of the PR, please take a look to Affected Dependencies section. If you have any doubt just reply or send me a slack message.

This PR has been a very good exercise to check what we need regarding documentation.

gmuraru commented on src/sympc/utils/utils.py:
```diff
@@ -77,9 +79,15 @@ def wrapper(
+            kwargs (Optional[Dict[Any, Dict[Any, Any]]]): Kwargs. Default to None.
+
+        Returns:
+            List[Any]
```
Q: Could you add here: Results from the parties?

gmuraru commented on src/sympc/utils/utils.py:
```diff
@@ -43,32 +48,29 @@ def parallel_execution(
-
-        :return: a function that runs in parallel at multiple parties or not
-        :rtype: a Callable that returns a list of results
+    """Wrap a function such tat it can be run in parallel at multiple parties.
```
TYPO: tat -- >that

gmuraru commented on src/sympc/utils/utils.py:
```diff
@@ -43,32 +48,29 @@ def parallel_execution(
+
+    Args:
+        fn (Callable): The function to run.
+        parties (Union[None, List[Any]]): Clients from syft. If this is set, then the f
```
Q: it is ok for the f to be on this line?

gmuraru commented on src/sympc/tensor/share_tensor.py:
```diff
@@ -283,24 +357,69 @@ def __eq__(self, other: Any) -> bool:
+            Any: Tensor with new view.
+
+        References:
+            https://pytorch.org/docs/stable/generated/torch.unsqueeze.html
```
TYPO: Should be ```view```

gmuraru commented on src/sympc/tensor/share_tensor.py:
```diff
@@ -283,24 +357,69 @@ def __eq__(self, other: Any) -> bool:
+            **kwargs: Keyword arguments passed to tensor.unsqueeze
+
+        Returns:
+            Any: Tensor unsqueezed.
```
Typo: ShareTensor

gmuraru commented on src/sympc/tensor/share_tensor.py:
```diff
@@ -283,24 +357,69 @@ def __eq__(self, other: Any) -> bool:
+            **kwargs: Keyword arguments passed to tensor.view.
+
+        Returns:
+            Any: Tensor with new view.
```
Typo: ShareTensor

gmuraru commented on src/sympc/tensor/share_tensor.py:
```diff
@@ -173,6 +201,14 @@ def mul(self, y: Union[int, float, torch.Tensor, "ShareTensor"]) -> "ShareTensor
+        """Apply the "xor" operation between "self" and "y".
+
+        Args:
+            y (Union[int, torch.Tensor, "ShareTensor"]): self @ y
```
TYPO: @ should be xor

gmuraru commented on src/sympc/tensor/share_tensor.py:
```diff
@@ -148,8 +170,11 @@ def sub(self, y: Union[int, float, torch.Tensor, "ShareTensor"]) -> "ShareTensor
-        :return: y - self
-        :rtype: ShareTensor
+        Args:
+            y (Union[int, float, torch.Tensor, "ShareTensor"]): self - y
```
TYPO: y - self

gmuraru commented:
Could you also add the option such that PRs would fail if ```docstring``` is not covered for at least 80%? (heads up @LaRiffle @arturomf94 @aanurraj)

LaRiffle commented on src/sympc/module/nn/conv.py:
```diff
@@ -78,7 +81,10 @@ def share_state_dict(self, state_dict: Dict[str, Any]) -> None:
+            state_dict (Dict[str, Any]): the state dict that would be shared.
+
+        Raises:
+            ValueError: If Kenel sizes mismatch "kernel_size_w" and "kernel_size_h"
```
some typo 
```suggestion
            ValueError: If kernel sizes mismatch "kernel_size_w" and "kernel_size_h"
```

LaRiffle commented on src/sympc/protocol/beaver/beaver.py:
```diff
@@ -103,6 +134,14 @@ def mul_store_add(
+    """Add the primitives required for the "mul" operation to the CryptoStore.
+
+    Arguments:
+        store: the CryptoStore
+        primitives: the list of primitives
+        a_shape: the shape of the first operand
+        b_shape: the shape of the second operand
```
```suggestion
        store (Dict[Any, Any]): the CryptoStore
        primitives (Iterable[Any]): the list of primitives
        a_shape (Tuple[int]): the shape of the first operand
        b_shape (Tuple[int]): the shape of the second operand
```

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -43,12 +47,18 @@
+        op (str): Operator eq or comp <=> DPF or DIF.
+
+    Returns:
+        noqa: DAR201
```
is this expected?

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -113,6 +124,18 @@ def fss_op(x1, x2, op="eq"):
 def mask_builder(
     session: Session, x1: ShareTensor, x2: ShareTensor, op: str
 ) -> ShareTensor:
+    """TODO: Add docstring.
```
```suggestion
    """Add the share of alpha (the mask) that is held in the crypto store to the 
    difference x1 - x2.
    
    As we aim at comparing x1 <= x2, we actually compare x1 - x2 <= 0 and we hide
    x1 - x2 with alpha that is a random mask.
```

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -128,6 +151,21 @@ def mask_builder(
 
 # share level
 def evaluate(session: Session, b, x_masked, op, dtype="long") -> ShareTensor:
+    """TODO: Add docstring.
```
```suggestion
    """Evaluate the FSS protocol on the masked and public input `x_masked`
```

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -172,29 +229,68 @@ def comp_evaluate(session: Session, b, x_masked, dtype=None) -> ShareTensor:
 
     @staticmethod
     def keygen(n_values=1):
+        """Sysecret DPF keygen.
```
```suggestion
        """Sycret DPF keygen.
```

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -172,29 +229,68 @@ def comp_evaluate(session: Session, b, x_masked, dtype=None) -> ShareTensor:
 
     @staticmethod
     def eval(b, x, k_b):
+        """TODO: Add docstring.
```
```suggestion
        """Sycret DPF eval.
```

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -172,29 +229,68 @@ def comp_evaluate(session: Session, b, x_masked, dtype=None) -> ShareTensor:
 
     @staticmethod
     def keygen(n_values=1):
+        """Sysecret DIF keygen.
```
```suggestion
        """Sycret DIF keygen.
```

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -172,29 +229,68 @@ def comp_evaluate(session: Session, b, x_masked, dtype=None) -> ShareTensor:
 
     @staticmethod
     def eval(b, x, k_b):
+        """TODO: Add docstring.
```
```suggestion
        """Sycret DIF eval.
```

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -228,19 +344,42 @@ def _generate_primitive(op: str, n_values: int) -> List[Any]:
 
 @register_primitive_generator("fss_eq")
 def generate_primitive(n_values: int) -> List[Any]:
+    """TODO: Add docstring.
```
```suggestion
    """Generate primitives for the FSS equality protocol.
```

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -228,19 +344,42 @@ def _generate_primitive(op: str, n_values: int) -> List[Any]:
 
 @register_primitive_generator("fss_comp")
 def generate_primitive(n_values: int) -> List[Any]:
+    """TODO: Add docstring.
```
```suggestion
    """Generate primitives for the FSS comparison protocol.
```

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -228,19 +344,42 @@ def _generate_primitive(op: str, n_values: int) -> List[Any]:
     primitives: Iterable[Any],
-) -> None:
+):
+    """TODO: Add docstring.
```
```suggestion
    """Add FSS primitives to the crypto store.
```

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -258,14 +397,29 @@ def add_primitive(
     store: Dict[Any, Any],
     primitives: Iterable[Any],
 ) -> None:
+    """TODO: Add docstring.
```
```suggestion
    """Add FSS primitives for equality to the crypto store.
```

LaRiffle commented on src/sympc/protocol/fss/fss.py:
```diff
@@ -258,14 +397,29 @@ def add_primitive(
-    primitives: Iterable[Any],
-) -> None:
+def add_primitive(store: Dict[Any, Any], primitives: Iterable[Any]):
+    """TODO: Add docstring.
```
```suggestion
    """Add FSS primitives for comparison to the crypto store.
```

LaRiffle commented on src/sympc/store/crypto_primitive_provider.py:
```diff
@@ -70,6 +78,11 @@ def _transfer_primitives_to_parties(
+        """Get the state of a CryptoProvider.
+
+        Returns:
+            str: CrytoProvider
```
```suggestion
            str: CryptoProvider
```

LaRiffle commented on src/sympc/store/crypto_store.py:
```diff
@@ -1,3 +1,5 @@
+"""CrypStore manages the needed crypto primitives."""
```
```suggestion
"""CryptoStore manages the needed crypto primitives."""
```

LaRiffle reviewed (commented):
Amazing work!!
Some suggestions below 👇

LaRiffle commented:
@jmaunon I have fixed already some typos & co! There are a few issues remaining with the linting but it's hard for me to fix them as the branch is on your repo. Could you have a look? I would very much like to merge your impressive work asap :)

jmaunon commented:
> @jmaunon I have fixed already some typos & co! There are a few issues remaining with the linting but it's hard for me to fix them as the branch is on your repo. Could you have a look? I would very much like to merge your impressive work asap :)

Thanks! This week I have been busy with the `docs-team` but this weekend should be fixed :)

gmuraru commented:
@jmaunon what do you think if we also have as mandatory an 80% total docs rate?
In this way, we should keep the repo up to at least 80% documentation.

jmaunon commented:
> @jmaunon what do you think if we also have as mandatory an 80% total docs rate?
> In this way, we should keep the repo up to at least 80% documentation.

Absolutely, I was doing that right now!

jmaunon commented:
@LaRiffle @gmuraru , from my side ir ready to merge
I have fixed a minor error, updated GA to fail if docstring-coverage < 80 and I have updated `pre-commit` to exclude `test` folder.

In any case, if later we see any problem/issue, you can send me a message via slack

gmuraru commented:
Great Work!

LaRiffle commented:
Excellent! Thanks again @jmaunon for the great work 😁
